torch/_tensor.py

```diff
@@ -107,6 +107,7 @@ def __deepcopy__(self, memo):
                     and self.device.type == torch._C._get_privateuse1_backend_name()
                 )
                 or (type(self) is not Tensor and self.data_ptr() == 0)
+                or type(self).__name__ == "DTensor"
```

Comment:
I think we don't want special case like this.

What's the actual failure you saw? If it's DTensor.clone() does not work we can probably just support that instead of special casing here.

Reply:
```
  File "/root/cery.69/pytorch/test/distributed/tensor/parallel/test_fsdp_2d_parallel.py", line 332, in test_2d_load_state_dict
    ref_state_dict = deepcopy(model_2d.state_dict())
  File "/usr/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.9/copy.py", line 296, in _reconstruct
    value = deepcopy(value, memo)
  File "/usr/lib/python3.9/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/root/cery.69/pytorch/torch/_tensor.py", line 182, in __deepcopy__
    new_tensor.set_(
  File "/root/cery.69/pytorch/torch/distributed/_tensor/api.py", line 290, in __torch_dispatch__
    return DTensor._op_dispatcher.dispatch(
  File "/root/cery.69/pytorch/torch/distributed/_tensor/dispatch.py", line 113, in dispatch
    self.sharding_propagator.propagate(op_info)
  File "/root/cery.69/pytorch/torch/distributed/_tensor/sharding_prop.py", line 164, in propagate
    output_sharding = self.propagate_op_sharding(op_info.schema)
  File "/root/cery.69/pytorch/torch/distributed/_tensor/sharding_prop.py", line 176, in propagate_op_sharding_non_cached
    out_tensor_meta = self._propagate_tensor_meta(op_schema)
  File "/root/cery.69/pytorch/torch/distributed/_tensor/sharding_prop.py", line 88, in _propagate_tensor_meta
    fake_out = op_schema.op(*fake_args, **fake_kwargs)
  File "/root/cery.69/pytorch/torch/_ops.py", line 572, in __call__
    return self_._op(*args, **(kwargs or {}))
  File "/root/cery.69/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/root/cery.69/pytorch/torch/_subclasses/fake_tensor.py", line 1436, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/root/cery.69/pytorch/torch/_subclasses/fake_tensor.py", line 1750, in dispatch
    r = func(*args, **kwargs)
  File "/root/cery.69/pytorch/torch/_ops.py", line 572, in __call__
    return self_._op(*args, **(kwargs or {}))
RuntimeError: Attempted to set the storage of a tensor on device "meta" to a storage on different device "cuda:1".  This is no longer allowed; the devices must match.
```

when the data_ptr is not null, the condition will go throught the next else branch.
and it copy the storage memo ? but it seems change the device meta to cuda

Reply:
@wanchaol hi,  how to solve this error? 
i want to implement a ```is_wrapped_tensor``` api for tensor , and use it to check whether tensor is a tensor_wrapper, do you think this change ok?

Reply:
Hey @JsBlueCat, sorry for the late reply! was busy last week and finally got a chance to repro this failures locally . I think in order to avoid special casing in `_tensor.py`, we need to implement the following op in DTensor:

```
aten.set_.source_Storage_storage_offset(Tensor(a!) self, Storage source, SymInt storage_offset, SymInt[] size, SymInt[] stride=[]) -> Tensor(a!)
```

We can probably implement this op as a custom handler in DTensor and directly calls into the `local_shard.set_`, but for the sharded DTensor case, we'll also need to calculate the "local shard stride", which might be a bit tricky. lmk if you want to give a try on this :)

Reply:
> local

okey，i will try to implement this. thanks!